OpenAI experiments (#87)
skuid/skuid-labs: experiments/dataSources/openAI/imageAdd/Readme.MD

# Ask DALL-E to generate sample images for your business records 
 
<img src="ImageSelection.png" width="500"></img>

> This sample page shows how the Dall-e API can be used to create images that are associated with data from your business records. The application allows the user to create a prompt that includes data from an account record. The prompt can be adjusted and then submitted to the image generation API.  4 images are returned.  The user may select one, and it will be saved to a Salesforce attachment record,  connected to the account record using Skuid's "File" custom sObject.  

Reviewer: ```suggestion
This sample page shows how the DALL-E API can be used to create images based on data from your business records. The application allows the user to create a prompt that includes data from an account record. The prompt can be adjusted and then submitted to the image generation API.  4 images are returned.  The user may select one, and it will be saved to a Salesforce attachment record, connected to the account record using Skuid's "File" custom sObject.  
```

## Instructions
- Page API:  V2
- Data sources: 
    - Salesforce data source.  (Account object,  Attachment object,  Skuid_Image__c object)

>     - REST data source named "OpenAi" connecting to your OpenAI instance per [instruction](openAI)

Reviewer: When using the page and data source as instructed I get this error when opening it in the Composer:

> Error retrieving metadata for model(s) associated with data source 'OpenAI'. Please check models' properties to ensure they are set correctly.
Error connecting to REST data source at URL "https://api.openai.com/v1/images/generations": Request to data source "OpenAI" failed: Upstream request error: 400 - Bad Request

Weirdly enough, the runtime page seems to work??? How can we avoid this error? If we can't avoid it, then we should note it in the instructions that it's expected.

Reviewer: Semi-related, and it's easier just to type it instead of putting a suggestion on the page XML: **we should remove the record type fields in this page's models.** Not all orgs have them enabled (like the one I tested this one), and on opening the page I get these errors:

> 1. A Skuid Model, 'Accounts', requested a Field with relationship name 'RecordType.Name', on the Account Object, but Skuid could not find a valid Field accessible through this relationship name. Please check that this Field actually exists on this Object (or on any related Objects). If it exists, then ensure that the user has permission to access it. If it does not exist, remove it from this Model.
> 2. A Skuid Model, 'Accounts', requested a Field with relationship name 'RecordTypeId', on the Account Object, but Skuid could not find a valid Field accessible through this relationship name. Please check that this Field actually exists on this Object (or on any related Objects). If it exists, then ensure that the user has permission to access it. If it does not exist, remove it from this Model.
> 3. A Skuid Model, 'ImageSave_Account', requested a Field with relationship name 'RecordType.Name', on the Account Object, but Skuid could not find a valid Field accessible through this relationship name. Please check that this Field actually exists on this Object (or on any related Objects). If it exists, then ensure that the user has permission to access it. If it does not exist, remove it from this Model.
> 4. A Skuid Model, 'ImageSave_Account', requested a Field with relationship name 'RecordTypeId', on the Account Object, but Skuid could not find a valid Field accessible through this relationship name. Please check that this Field actually exists on this Object (or on any related Objects). If it exists, then ensure that the user has permission to access it. If it does not exist, remove it from this Model.

Reviewer: Okay, runtime isn't a total success. I get a 502 error when I click "Save image to account" and the page freezes. Any ideas on how to fix this issue or instructions on how to avoid it? Context: I'm in an NLX site connecting to a Salesforce org through OAuth—which I assume should work as expected with no additional steps.

<img width="522" alt="Screenshot 2023-06-09 at 1 10 51 PM" src="https://github.com/skuid/skuid-labs/assets/7629304/0289a88f-0ea2-4f06-a3e9-0f9ebbc955d1">

- Design system: None 
- Page XML:  [Copy the XML from this page](AI_Image_Generation.xml?raw=true), or save it as an XML file, and upload it as a new page in Skuid's Pages.

## Notes

> ### Models:  
> 
> What are the models in this demo doing? 

Reviewer: ```suggestion
### Models
```
